core/src/main/java/com/radixdlt/consensus/LedgerHashes.java

```diff
@@ -0,0 +1,160 @@
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.consensus;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.hash.HashCode;
+import com.radixdlt.crypto.HashUtils;
+import com.radixdlt.serialization.DsonOutput;
+import com.radixdlt.serialization.SerializerConstants;
+import com.radixdlt.serialization.SerializerDummy;
+import com.radixdlt.serialization.SerializerId2;
+import java.util.Objects;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+@SerializerId2("consensus.ledger_hashes")
+public final class LedgerHashes {
+  @JsonProperty(SerializerConstants.SERIALIZER_NAME)
+  @DsonOutput(value = {DsonOutput.Output.API, DsonOutput.Output.WIRE, DsonOutput.Output.PERSIST})
+  SerializerDummy serializer = SerializerDummy.DUMMY;
+
+  @JsonProperty("state_root")
+  // TODO: restore `Output.ALL` after fixing non-determinism bugs
+  // We need to come up with a few tests that expose potential non-determinism bugs, and then fix
+  // them. Before that happens, it could be risky to include the state hash in the DSON-based
+  // hashing (used e.g. in `QuorumCertificate` to get a hash of `VoteData`, leading to nodes not
+  // being able to agree on the same hash).
+  @DsonOutput(value = DsonOutput.Output.HASH, include = false)
+  private final HashCode stateRoot;
+
+  @JsonProperty("transaction_root")
+  @DsonOutput(DsonOutput.Output.ALL)
+  private final HashCode transactionRoot;
+
+  @JsonProperty("receipt_root")
+  // TODO: restore `Output.ALL` after fixing non-determinism bugs
+  // The test `REv2ConsensusLedgerRecoveryTest.recovery_should_work_when_consensus_is_behind_ledger`
+  // fails when the receipt root is taken into account, which suggests that nodes can end up with
+  // different receipts in some special cases. We need to investigate and fix it.
```

Comment:
I investigated this, and indeed there are differing receipts of exactly the same transaction!
(in that test, it was a genesis transaction, but I don't think it's a prerequisite)

Luckily, from the diff, it seems we _only_ have a problem with order of items in a `resource_changes` vec:

![image](https://user-images.githubusercontent.com/119413677/221938353-b866b681-5eaf-40f8-bb88-7689a86b8b88.png)

And indeed, here is where it comes from:
![image](https://user-images.githubusercontent.com/119413677/221939983-3143a8ca-96cd-4c88-bc75-5ae5dfc65ee3.png)

A HashMap of HashMaps, what can go wrong? :P

Let me fix this.

Reply:
_(it will be a fix in scrypto lib, so this PR still has to go in as-is)_